cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/DefaultProgramWorkflowRunner.java

```diff
@@ -149,6 +155,13 @@ private void runAndWait(ProgramRunner programRunner, Program program, ProgramOpt
       Closeables.closeQuietly(closeable);
       throw t;
     }
+
```

Comment:
Since the `MapReduceProgramRunner` or `SparkProgramRunner` no longer persists program states itself, we need to add the `ProgramStateChangeListener` to the MapReduce or Spark controller inside a workflow node.